client/ayon_core/pipeline/workfile/workfile_template_builder.py

```diff
@@ -1102,6 +1135,41 @@ def set_plugin_shared_populate_data(self, key, value):
         plugin_data[key] = value
         self.builder.set_shared_populate_data(self.identifier, plugin_data)
 
+    def register_on_finished_callback(
+            self, placeholder, callback, order=None
+    ):
+        self.register_callback(
+            placeholder,
+            topic="template.finished",
+            callback=callback,
+            order=order
+        )
+
+    def register_on_depth_processed_callback(
+            self, placeholder, callback, order=0
+    ):
+        self.register_callback(
+            placeholder,
+            topic="template.depth_processed",
+            callback=callback,
+            order=order
+        )
+
+    def register_callback(self, placeholder, topic, callback, order=None):
+
+        if order is None:
+            # Match placeholder order by default
+            order = placeholder.order
+
+        # We must persist the callback over time otherwise it will be removed
+        # by the event system as a valid function reference. We do that here
+        # always just so it's easier to develop plugins where callbacks might
+        # be partials or lambdas
+        placeholder.data.setdefault("callbacks", []).append(callback)
```

Comment:
Callbacks should not be stored to placeholder data.

Reply:
Fixed with https://github.com/ynput/ayon-core/pull/426/commits/edee279f15fa16b549db68e3adb6aba14fc2f541